src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java

```diff
@@ -902,6 +896,45 @@ private void setTypeDefinition(DTV dtv) {
             }
         }
 
+
+        /**
+         * Generates the SQL datatype to use for Java date-based values. This
+         * setting can be controlled by setting the "datetimeParameterType" connection
+         * string. It defaults to "datetime2" for SQL Server 2018+ and always
+         * uses "datetime" for older SQL Server installations.
+         */
+        String getDatetimeDataType(SQLServerConnection con, Integer scale) {
+            String datatype;
+
+            if (con.isKatmaiOrLater()) {
+                switch (con.getDatetimeParameterType()){
+                    case "datetime2":
+                        datatype = SSType.DATETIME2.toString();
+                        if (scale != null){
+                            datatype += "(" + scale + ")";
+                        }
+                        return datatype;
+                    case "datetimeoffset":
+                        datatype = SSType.DATETIMEOFFSET.toString();
+                        if (scale != null){
+                            datatype += "(" + scale + ")";
+                        }
+                        return datatype;
+                    case "datetime":
+                    default:
+                        return SSType.DATETIME.toString();
+
+                }
+            }
+
+            /*
+                For older versions of SQL server and for some reason the datetimeParameterType
+                connection property cannot be determined, we fall back to the "datetime"
+                format.
+            */
+            return SSType.DATETIME.toString();
+        }
```

Comment:
I'm wondering how this should be handled, if one explicitly sets data2 but it's pre SQL Server 2008. Should this silently change to datetime or error or just log a warning? Don't make any changes here yet I'll have to check with our product manager next week to see which he prefers.

Reply:
For pre-Katmai, I wouldn't do anything here. Maybe there should be a warning in the log if you try to set the value saying the parameter is unsupported.

For Katmai, I wasn't sure how you generally approach things like this. I was going to fault tolerance and trying to minimize any backwards compatibility issues. I think it would make some sense to throw a hard error if an invalid value is supplied, but a warning message certainly makes sense.